### PR TITLE
Various Enhancements

### PR DESCRIPTION
**Pull request by mavenlin**
I've used this project for some harder cases and figure out some edits needed. Here's part of them

1. support passing kwargs for `mean` / `sum`.
2. fix `view`'s implementation when a single `tuple` instead of multiple `int` is passed.
3. better support attention with `jax.nn.dot_product_attention`.
4. add various frequently used functions, e.g. `sin/cos`, `unsqueeze`, `to`, `contiguous`, `device` etc.

### CODE REVIEW

File: torch2jax/__init__.py [hunk 1]
```diff
@@ -101,6 +101,8 @@ def dtype(self) -> torch.dtype: return j2t_dtype(self.value.dtype)
   def ndim(self) -> int: return len(self.value.shape)
   @property
   def shape(self): return self.value.shape
+  @property
+  def device(self): return torch.device("cpu")
```
Comment (samuela) [inline torch2jax/__init__.py:105]: This is fishy to me. Why should we return "cpu" here as opposed to "cuda:0" or anything else?

File: torch2jax/__init__.py [hunk 2]
```diff
@@ -157,11 +160,22 @@ def pow(*args, **kwargs): return torch.pow(*args, **kwargs)
...
+  def cos(*args, **kwargs): return torch.cos(*args, **kwargs)
+  def unsqueeze(self, dim): return Torchish(jnp.expand_dims(self.value, axis=dim))
+  def float(self): return Torchish(jnp.astype(self.value, jnp.float32))
+  def to(self, *args, **kwargs): return self  # ignore device movement, jax manages its own placement.
+  def contiguous(self): return self
   # fmt: on
 
+  def view(self, *shape):
+    if len(shape) == 1 and isinstance(shape[0], Sequence):
+      shape = shape[0]
```
Comment (samuela) [inline torch2jax/__init__.py:174]: let's add a test case for this. you can add it here: https://github.com/samuela/torch2jax/blob/7132ab7d7354a12afed82902581180de143aabf3/tests/test_core.py#L118-L122

File: torch2jax/__init__.py [hunk 3]
```diff
@@ -242,15 +256,39 @@ def fn(*args, **kwargs):
...
 auto_implements(torch.transpose, jnp.swapaxes)
+auto_implements(torch.rsqrt, lambda x: jnp.reciprocal(jnp.sqrt(x)))
+
+
+@implements(torch.nn.functional.silu)
+def silu(x, *args, **kwargs):
+  x = _v(x)
+  if "inplace" in kwargs and kwargs["inplace"]:
+    assert not kwargs["inplace"], "jax doesn't support inplace"
+  return jax.nn.silu(x)
```
Comment (samuela) [inline torch2jax/__init__.py:276]: please place this in alphabetical order in the torch.nn.functional section below: https://github.com/samuela/torch2jax/blob/7132ab7d7354a12afed82902581180de143aabf3/torch2jax/__init__.py#L519

File: torch2jax/__init__.py [hunk 4]
```diff
@@ -242,15 +256,39 @@ def fn(*args, **kwargs):
...
 auto_implements(torch.permute, jnp.transpose, dont_coerce_argnums=(1, 2))  # TODO: do we need argnum 2?
 auto_implements(torch.pow, jnp.power)
+auto_implements(torch.sin, jnp.sin)
+auto_implements(torch.cos, jnp.cos)
 auto_implements(torch.sigmoid, jax.nn.sigmoid)
 auto_implements(torch.sqrt, jnp.sqrt)
-auto_implements(torch.sum, jnp.sum)
 auto_implements(torch.tanh, jnp.tanh)
 auto_implements(torch.transpose, jnp.swapaxes)
+auto_implements(torch.rsqrt, lambda x: jnp.reciprocal(jnp.sqrt(x)))
```
Comment (samuela) [inline torch2jax/__init__.py:268]: please use https://docs.jax.dev/en/latest/_autosummary/jax.lax.rsqrt.html

File: torch2jax/__init__.py [hunk 5]
```diff
@@ -242,15 +256,39 @@ def fn(*args, **kwargs):
 auto_implements(torch.add, jnp.add)
 auto_implements(torch.exp, jnp.exp)
 auto_implements(torch.nn.functional.gelu, jax.nn.gelu)
-auto_implements(torch.mean, jnp.mean)
 auto_implements(torch.mul, jnp.multiply)
 auto_implements(torch.permute, jnp.transpose, dont_coerce_argnums=(1, 2))  # TODO: do we need argnum 2?
 auto_implements(torch.pow, jnp.power)
+auto_implements(torch.sin, jnp.sin)
+auto_implements(torch.cos, jnp.cos)
```
Comment (samuela) [inline torch2jax/__init__.py:263]: let's keep this list in alphabetical order

File: torch2jax/__init__.py [hunk 6]
```diff
@@ -242,15 +256,39 @@ def fn(*args, **kwargs):
...
 auto_implements(torch.sigmoid, jax.nn.sigmoid)
 auto_implements(torch.sqrt, jnp.sqrt)
-auto_implements(torch.sum, jnp.sum)
 auto_implements(torch.tanh, jnp.tanh)
 auto_implements(torch.transpose, jnp.swapaxes)
+auto_implements(torch.rsqrt, lambda x: jnp.reciprocal(jnp.sqrt(x)))
+
+
+@implements(torch.nn.functional.silu)
+def silu(x, *args, **kwargs):
```
Comment (samuela) [inline torch2jax/__init__.py:272]:
```suggestion
def silu(input, inplace=True):
```
to match the pytorch arguments

File: torch2jax/__init__.py [hunk 7]
```diff
@@ -242,15 +256,39 @@ def fn(*args, **kwargs):
...
+  return jax.nn.silu(x)
+
+
+@implements(torch.device, Torchishify_output=False)
+def torch_device(type):
+  return torch.device(type)
+
+
+@implements(torch.mean)
+def mean(tensor, dim=None, keepdim=False, dtype=None):
```
Comment (samuela) [inline torch2jax/__init__.py:285]:
```suggestion
def mean(input, dim=None, keepdim=False, dtype=None):
```

so that `torch.mean(input=foo)` works, matching the pytorch definition

File: torch2jax/__init__.py [hunk 8]
```diff
@@ -242,15 +256,39 @@ def fn(*args, **kwargs):
...
+@implements(torch.nn.functional.silu)
+def silu(x, *args, **kwargs):
+  x = _v(x)
+  if "inplace" in kwargs and kwargs["inplace"]:
+    assert not kwargs["inplace"], "jax doesn't support inplace"
+  return jax.nn.silu(x)
+
+
+@implements(torch.device, Torchishify_output=False)
+def torch_device(type):
```
Comment (samuela) [inline torch2jax/__init__.py:280]:
```suggestion
def device(type):
```

Comment (samuela) [reply torch2jax/__init__.py:280]: what would a test case that requires this look like? doesn't this just auto-fallback to the normal torch.device implementation?

File: torch2jax/__init__.py [hunk 9]
```diff
@@ -242,15 +256,39 @@ def fn(*args, **kwargs):
...
+  return torch.device(type)
+
+
+@implements(torch.mean)
+def mean(tensor, dim=None, keepdim=False, dtype=None):
+  return jnp.mean(_v(tensor), axis=dim, keepdims=keepdim, dtype=t2j_dtype(dtype or tensor.dtype))
+
+
+@implements(torch.sum)
+def sum(tensor, dim=None, keepdim=False, dtype=None):
```
Comment (samuela) [inline torch2jax/__init__.py:290]:
```suggestion
def sum(input, dim=None, keepdim=False, dtype=None):
```

Comment (samuela) [reply torch2jax/__init__.py:290]: please keep these functions in alphabetical order

**Review by samuela — COMMENTED** [review]
Thanks for submitting this @mavenlin ! Let's add tests for the methods that are added/changed. Check out https://github.com/samuela/torch2jax/blob/main/tests/test_core.py and https://github.com/samuela/torch2jax/blob/main/tests/test_torch_nn.py for some examples. Feel free to reach out if you have any questions!

File: torch2jax/__init__.py (hunk 2 above)
Comment (mavenlin) [reply torch2jax/__init__.py:174]: Sure, what is the local configuration when you run the test? I figure that the test wouldn't run unless I turn the GPU off. Complaining about `tensor.numpy()` should be `tensor.cpu().numpy()`. Do we want to add an ci script to github?

File: torch2jax/__init__.py (hunk 7 above)
Comment (mavenlin) [reply torch2jax/__init__.py:285]: you're right

File: torch2jax/__init__.py (hunk 8 above)
Comment (mavenlin) [reply torch2jax/__init__.py:280]: Yes, this autofalls back, due to the above `torch.device("cpu")` call would be intercepted by `__torch_function__` mechanism, that's why the need for this function. I think we just need to keep the original `torch.device` and can not replace with `jax.Device` here because the code that consume this object are python code that are not part of `__torch_function__`.

File: tests/test_torch_nn.py [hunk 10]
```diff
@@ -350,13 +350,29 @@ def test_torch_nn_functional_prelu():
...
 def test_torch_nn_functional_scaled_dot_product_attention():
   t2j_function_test(lambda x, y: x @ y, [(2, 3, 5), (5, 7)], atol=1e-6)
 
   sdpa = torch.nn.functional.scaled_dot_product_attention
+
+  # default
   t2j_function_test(sdpa, [(5, 3, 7), (5, 2, 7), (5, 2, 7)], atol=1e-6)
   t2j_function_test(sdpa, [(5, 7, 11), (5, 7, 11), (5, 7, 11)], atol=1e-6)
 
+  # default + attn_mask
```
Comment (samuela) [inline tests/test_torch_nn.py:367]: can you please also add a test for the case when `attn_mask` has dtype bool? i think that should cover all branches

File: torch2jax/__init__.py [hunk 11]
```diff
@@ -95,6 +94,13 @@ def __torch_function__(cls, func, types, args=(), kwargs=None):
     raise NotImplementedError(f"Torchish.__torch_function__: {func}")
 
   # fmt: off
+  @property
+  def device(self):
+    try:  # Attempt to get the device from the value's device.
+      return torch.device(self.value.device.type.replace("gpu", "cuda"))
+    except Exception:  # otherwise, assume CPU, e.g. abstrace arrays, tpu arrays
+      return torch.device("cpu")
```
Comment (samuela) [inline torch2jax/__init__.py:102]: hmm, i think you've successfully sold me on making everything "cpu" here... why change to this? In general, dealing with devices across the jax/torch boundary seems like messy business, so i'm a bit more hesitant when picking this api. i especially want to avoid complexity. open to ideas, just want to make sure we get it right

now that i write this comment, would it also make sense to have `torch.device(<anything>)` return "cpu"? that way everything is consistent

if going this route, why is the `.replace("gpu", "cuda")` necessary?

**Review by samuela — COMMENTED** [review]
> I think we can safely set it to cpu for all cases because for jax there is no explicitly setting device during computation, the device placement is managed automatically. The __torch_function__ mechanism will convert the function call to jax which under the hook will derive the device for all subsequent computations.

i see. i think this makes sense. this is one of the trickier parts of the api that we have thus far just avoided, but it's good that we address it. see my line comment for more

> With the above said, I have some concern on how the inplace is handled in the code, as jax doesn't support inplace at all. Currently I see that it replaces the value field in the Torchish object. However, it is complicated if we take grad or jit on it.

good question! we actually handle this via interior mutation on `Torchish`s, specifically on the `value` attribute. So from the Torchish perspective there is mutation, but under the hood everything is happening with immutable JAX objects. See https://github.com/samuela/torch2jax/blob/7132ab7d7354a12afed82902581180de143aabf3/tests/test_core.py#L195-L207 for an example test case.

thanks for adding these tests! i think we're close now. 2 things left afaict:
- test for boolean `attn_mask`s
- decision on and implementation of the right `device` api. i'm happy to go with an api where `<anytensor>.device == torch.device(<anything>) == "cpu"`. see comment below for more discussion

**Comment by mavenlin** [comment]
> good question! we actually handle this via interior mutation on `Torchish`s, specifically on the `value` attribute. So from the Torchish perspective there is mutation, but under the hood everything is happening with immutable JAX objects. See

Yes it does do the job if we're not doing differentiation. However, the torch code would err on backprop detecting an inplace modification, while Jax would still backprop down the new value. We may want to take note of this.

> thanks for adding these tests! i think we're close now. 2 things left afaict:
> 
> * test for boolean `attn_mask`s

Will do, I figure out how to minimally change the test template.

> * decision on and implementation of the right `device` api. i'm happy to go with an api where `<anytensor>.device == torch.device(<anything>) == "cpu"`. see comment below for more discussion

I also think returning all "cpu" would be the simplest and most consistent way. Keeping some "gpu" could end up with torch ops complaining device mismatch for its inputs.

File: torch2jax/__init__.py (hunk 11 above)
Comment (mavenlin) [reply torch2jax/__init__.py:102]: Let's unify with "cpu" then. So far I can't think of an example how this could fail.

File: tests/utils.py [hunk 12]
```diff
@@ -25,12 +27,16 @@ def test_t2j_array():
...
   for test_rng in random.split(rng, num_tests):
     # This is a thunk since methods like torch.nn.functional.batch_norm mutate the inputs and that affects subsequent
     # tests. We construct fresh values each time as a mitigation.
     n_inputs = len(input_shapes)
-    inputs = lambda: [random.normal(rng, shape) for rng, shape in zip(random.split(test_rng, n_inputs), input_shapes)]
+    if not isinstance(samplers, Sequence):
+      samplers = [samplers] * n_inputs
+    inputs = lambda: [
+      sampler(rng, shape=shape) for rng, shape, sampler in zip(random.split(test_rng, n_inputs), input_shapes, samplers)
+    ]
```
Comment (samuela) [inline tests/utils.py:39]:
```suggestion
def t2j_function_test(f, input_shapes, samplers=None, rng=random.PRNGKey(123), num_tests=5, **assert_kwargs):
  for test_rng in random.split(rng, num_tests):
    # This is a thunk since methods like torch.nn.functional.batch_norm mutate the inputs and that affects subsequent
    # tests. We construct fresh values each time as a mitigation.
    n_inputs = len(input_shapes)
    samplers = [random.normal] * n_inputs if samplers is None else samplers
    inputs = lambda: [
      sampler(rng, shape=shape) for rng, shape, sampler in zip(random.split(test_rng, n_inputs), input_shapes, samplers)
    ]
```
not sure if i got exactly the right diff, but this design is a preferable api imho

File: torch2jax/__init__.py [hunk 13]
```diff
@@ -297,6 +313,15 @@ def cat(tensors, dim=0):
   return jnp.concatenate([_v(x) for x in tensors], axis=dim)
 
 
+@implements(torch.device, Torchishify_output=False)
+def device(device):
+  # device doesn't matter to jax at all, because jax has its own implicit device
+  # management, the user has no mechanism to do something like `to(device)`.
+  # Therefore, we always return a CPU device, which makes the torch side to be
+  # always consistent.
```
Comment (samuela) [inline torch2jax/__init__.py:321]: thanks for adding this!

**Comment by mavenlin** [comment]
> hmm interesting. we currently have a test that it agrees with `torch.func.grad(inplace_fn)` and that works as expected. maybe `tensor.backward()` works differently? it would be funny if we had better inplace gradient support than pytorch itself

Didn't notice that the gradient is automatically tested. Nevermind then if it is working, my impression is that it throws an error but back in before 2021. I guess torch has did something to fix it.

File: torch2jax/__init__.py [hunk 14]
```diff
@@ -936,26 +983,28 @@ def prelu(input: Torchish, weight: Torchish):
...
+      mask = attn_mask
+    elif jnp.issubdtype(attn_mask.dtype, jnp.floating):
+      bias = attn_mask
+    else:
+      raise ValueError(f"Unsupported attn_mask dtype: {attn_mask.dtype}. Expected bool or float.")
+  output = jax.nn.dot_product_attention(Q, K, V, scale=scale, mask=mask, bias=bias, is_causal=is_causal)
+  output = jnp.swapaxes(output, -2, -3)
+  if mask is not None:
+    # (batch(optional), num_heads(optional), seq_len, 1)
+    output_mask = jnp.expand_dims(jnp.any(mask, axis=-1), axis=-1)
```
Comment (samuela) [inline torch2jax/__init__.py:1005]:
```suggestion
    output_mask = jnp.any(mask, axis=-1, keepdims=True)
```
why not keepdims?

File: torch2jax/__init__.py [hunk 15]
```diff
@@ -936,26 +983,28 @@ def prelu(input: Torchish, weight: Torchish):
...
+    attn_mask = _v(attn_mask)
+    if jnp.issubdtype(attn_mask.dtype, jnp.bool_):
+      mask = attn_mask
+    elif jnp.issubdtype(attn_mask.dtype, jnp.floating):
+      bias = attn_mask
+    else:
+      raise ValueError(f"Unsupported attn_mask dtype: {attn_mask.dtype}. Expected bool or float.")
+  output = jax.nn.dot_product_attention(Q, K, V, scale=scale, mask=mask, bias=bias, is_causal=is_causal)
+  output = jnp.swapaxes(output, -2, -3)
+  if mask is not None:
```
[the remark this answers is not in the document]
Comment (samuela) [reply torch2jax/__init__.py:1003]: oooh good catch. in that case, can we please add a test case for this and a code comment with your explanation of this situation?

[the remark this answers is not in the document]
Comment (mavenlin) [reply torch2jax/__init__.py:1003]: done

File: torch2jax/__init__.py (hunk 14 above)
Comment (mavenlin) [reply torch2jax/__init__.py]: Good catch
